key_management/src/keystore.rs

```diff
@@ -412,50 +409,66 @@ impl KeyStore {
 impl EncryptedKeyStore {
     fn derive_key(
         passphrase: &str,
-        prev_salt: Option<Vec<u8>>,
-    ) -> Result<(pwhash::Salt, Arc<secretbox::Key>), EncryptedKeyStoreError> {
+        prev_salt: Option<[u8; RECOMMENDED_SALT_LEN]>,
+    ) -> anyhow::Result<([u8; RECOMMENDED_SALT_LEN], Vec<u8>)> {
         let salt = match prev_salt {
-            Some(prev_salt) => match pwhash::Salt::from_slice(&prev_salt) {
-                Some(salt) => salt,
-                None => return Err(EncryptedKeyStoreError::ConfigurationError),
-            },
-            None => pwhash::gen_salt(),
+            Some(prev_salt) => prev_salt,
+            None => {
+                let mut salt = [0; RECOMMENDED_SALT_LEN];
+                OsRng.fill_bytes(&mut salt);
+                salt
+            }
         };
-        let mut key = secretbox::Key([0; secretbox::KEYBYTES]);
-
-        let secretbox::Key(ref mut kb) = key;
-        pwhash::derive_key(
-            kb,
-            passphrase.as_bytes(),
-            &salt,
-            pwhash::OPSLIMIT_INTERACTIVE,
-            pwhash::MEMLIMIT_INTERACTIVE,
-        )
-        .unwrap();
-
-        Ok((salt, Arc::new(key)))
-    }
 
-    fn encrypt(encryption_key: Arc<secretbox::Key>, msg: &[u8]) -> Vec<u8> {
-        let nonce = secretbox::gen_nonce();
-
-        let mut ciphertext = secretbox::seal(msg, &nonce, &encryption_key);
-        ciphertext.append(&mut nonce.as_ref().to_vec());
-        ciphertext
+        let hasher = {
+            let mut param_builder = ParamsBuilder::new();
+            // #define crypto_pwhash_argon2id_MEMLIMIT_INTERACTIVE 67108864U
+            param_builder.m_cost(67108864 / 1024).unwrap();
```

Comment:
Can we use `?` here? Is `argon2::Error` well-behaved?

Reply:
Fixed. Need to move it out and map_err tho